crates/flight_datafusion/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+            schema,
+            table_reference,
+        })
+    }
+
+    #[allow(clippy::needless_pass_by_value)]
+    fn get_schema<'a>(
+        client: Arc<FlightClient>,
+        table_reference: impl Into<TableReference<'a>>,
+    ) -> Result<SchemaRef> {
+        let mut client = client.deref().clone();
+        tokio::task::block_in_place(move || {
+            Handle::current().block_on(async {
+                let mut stream = client
+                    .query(format!("SELECT * FROM {} limit 1", table_reference.into()).as_str())
+                    .await
+                    .map_err(|error| Error::Flight { source: error })?;
+
+                if stream.next().await.is_some() {
+                    if let Some(schema) = stream.schema() {
+                        Ok(Arc::clone(schema))
+                    } else {
+                        Err(Error::NoSchema {})
+                    }
+                } else {
+                    Err(Error::NoSchema {})
+                }
+            })
+        })
+    }
+
+    fn create_physical_plan(
+        &self,
+        projections: Option<&Vec<usize>>,
+        schema: &SchemaRef,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(FlightSQLExec::new(
+            projections,
+            schema,
+            &self.table_reference,
+            Arc::clone(&self.client),
+            filters,
+            limit,
+        )?))
+    }
+}
+
+#[async_trait]
+impl TableProvider for FlightSQLTable {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> DataFusionResult<Vec<TableProviderFilterPushDown>> {
+        let mut filter_push_down = vec![];
+        for filter in filters {
+            match expr::to_sql(filter) {
+                Ok(_) => filter_push_down.push(TableProviderFilterPushDown::Exact),
+                Err(_) => filter_push_down.push(TableProviderFilterPushDown::Unsupported),
+            }
+        }
+
+        Ok(filter_push_down)
+    }
+
+    async fn scan(
+        &self,
+        _state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        return self.create_physical_plan(projection, &self.schema(), filters, limit);
+    }
+}
+
+#[derive(Clone)]
+struct FlightSQLExec {
+    projected_schema: SchemaRef,
+    table_reference: OwnedTableReference,
+    client: Arc<FlightClient>,
+    filters: Vec<Expr>,
+    limit: Option<usize>,
+}
+
+impl FlightSQLExec {
+    fn new(
+        projections: Option<&Vec<usize>>,
+        schema: &SchemaRef,
+        table_reference: &OwnedTableReference,
+        client: Arc<FlightClient>,
```

Comment:
There is no point wrapping a FlightClient in an Arc - because as you've discovered, we need to clone it to be able to query. I would just remove the Arc<>

Reply:
The real way to fix this would be using a connection pooler

Reply:
sounds good, can be done when refactoring them.